src/gluonts/dataset/parallelized_loader.py

```diff
@@ -0,0 +1,626 @@
+                )
+                if not isinstance(
+                    self._worker_pool, multiprocessing.pool.ThreadPool
+                ):
+                    msg += """- Insufficient shared_memory if `timeout` is large enough.
+                Please consider reduce `num_workers` or increase shared_memory in system.
+                """
+                print(msg)
+                raise
+            except Exception:
+                self._worker_pool.terminate()
+                raise
+
+    def __iter__(self):
+        while True:
+            next_batch = next(self)
+            if len(next_batch) == 0:
+                return
+            yield next_batch
+
+
+# TODO: think about how a multiprocessing.Manager() would complement this implementation
+class ParallelDataLoader(object):
+    """Loads data from a dataset and returns mini-batches of data.
+    Parameters
+    ----------
+    dataset
+        The dataset from which to load data.
+    transformation
+        A transformation to apply to each entry in the dataset.
+    batch_size
+        Size of mini-batch.
+    ctx
+        MXNet context to use to store data.
+    dtype
+        Floating point type to use.
+    shuffle
+        Whether to shuffle the samples.
+    sampler
+        The sampler to use. Either specify sampler or shuffle, not both.
+    num_workers
+        The number of multiprocessing workers to use for data preprocessing.
+    prefetch
+        The number of prefetching batches only works if `num_workers` > 0.
+        If `prefetch` > 0, it allow worker process to prefetch certain batches before
+        acquiring data from iterators.
+        Note that using large prefetching batch will provide smoother bootstrapping performance,
+        but will consume more shared_memory. Using smaller number may forfeit the purpose of using
+        multiple worker processes, try reduce `num_workers` in this case.
+        By default it defaults to `num_workers * 2`.
+    """
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        transformation: Transformation,
+        cyclic: bool,
+        is_train: bool,
+        batch_size: int,
+        shuffle: bool = False,
+        batchify_fn: Callable = None,
+        ctx: mx.Context = None,
+        dtype: DType = np.float32,
+        prefetch: int = None,
+        num_workers: int = 0,
+    ):
+        self.cyclic = (
+            cyclic  # indicates that we want to cycle through the dataset
+        )
+        self.dtype = dtype
+        self.is_train = is_train
+        self.transformation = transformation
+        self.ctx = ctx
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+
+        self.num_workers = num_workers if num_workers >= 0 else 0
+        self.worker_pool = None
+        self.prefetch = max(
+            0, int(prefetch) if prefetch is not None else 2 * self.num_workers
+        )
+
+        self.dataset = dataset
+        self.dataset_len = len(list(dataset))
```

Comment:
This loads the whole dataset into memory. 
Let's maybe remove this and switch to the modulo approach, where we don't need the dataset_len.

Reply:
yeah that was just there for debugging originally, I removed it.